lib/diyepw/diyepw/meteorology.py

```diff
@@ -0,0 +1,263 @@
+        if isinstance(val, list):
+            # Lists replace the items in a column one-to-one, so we have to ensure that the passed list contains
+            # exactly the same number of elements as our observations have rows
+            if len(val) != len(self._observations):
+                raise Exception("""
+                    This meteorological year has {my_num} observations, but you passed a 
+                    list of {your_num} replacement values
+                """.format(my_num=len(self._observations), your_num=len(val))
+                )
+
+        # An assignment works here regardless of whether val is a list or a single item. If it's a list, the whole
+        # column will be replaced with the values from the list. If it's a single item, that item will be duplicated
+        # to fill the entire column
+        self._observations[column_name] = val
+
+    # adapted from https://github.com/SSESLab/laf/blob/master/LAF.py
+    ####################################################################################################################
+    # Write new EPW file
+    ####################################################################################################################
+    def write_epw(self, save_path):
+        first_observation = {k:v[0] for (k, v) in self._observations[0:1].items()}
+
+        location_header = ",".join([str(i) for i in [
+            'LOCATION', self._city, self._state, self._country, 'customized weather file', self._station_number,
+            self._latitude, self._longitude, self._timezone_gmt_offset, self._elevation
+        ]])
+        first_observation_date = datetime.date(
+            year=int(first_observation["year"]),
+            month=int(first_observation["month"]),
+            day=int(first_observation["day"])
+        )
+        first_day_of_week = calendar.day_name[first_observation_date.weekday()]
+
+        with open(save_path, 'w') as epw_file:
+            epw_file.write("\n".join([
+                location_header,
+                "\n".join(self._headers[1:5]),
+                self._comment,
+
+                'COMMENTS 2, TMY3 data from energyplus.net/weather supplemented with NOAA ISD Lite data from ' +
+                'https://www1.ncdc.noaa.gov/pub/data/noaa/isd-lite/ for an actual meteorological year (AMY)',
+
+                'DATA PERIODS,1,1,Data,' + first_day_of_week + ', 1/1, 12/31'
+            ]) + "\n")
+            epw_file.write(self._observations.to_csv(header=False))
+
+    # adapted from https://github.com/SSESLab/laf/blob/master/LAF.py
+    @staticmethod
+    def from_tmy3_file(file_path:str):
+        """
+        Create an instance of this class based on a tmy3 file
+
+        :param file_path: Path to a TMY file. For the definition of a tmy3 file,
+        see https://www.nrel.gov/docs/fy08osti/43156.pdf
+        :return:
+        """
+        instance = Meteorology()
+        tmy3_file = open(file_path)
```

Comment:
use `with open(...)` in case of error before close to prevent file locking

Reply:
Done